src/dimse/wadoRs.ts

```diff
@@ -1,72 +1,287 @@
 import { ConfParams, config } from '../utils/config';
 import { LoggerSingleton } from '../utils/logger';
-import { fileExists } from '../utils/fileHelper';
 import { waitOrFetchData } from './fetchData';
 import { compressFile } from './compressFile';
-import dicomParser from 'dicom-parser';
-import crypto from 'crypto';
+import { doFind } from '../dimse/findData';
 import path from 'path';
-import fs from 'fs';
+import fs from 'fs/promises';
 import { QUERY_LEVEL } from './querLevel';
+import deepmerge from 'deepmerge';
+import dicomParser from 'dicom-parser';
+import combineMerge from '../utils/combineMerge';
+import { fileExists, setWadoInProgress } from '../utils/fileHelper';
+import { execFile as exFile } from 'child_process';
+import util from 'util';
+
+const execFile = util.promisify(exFile);
+
+export type DataFormat = 'pixeldata' | 'bulkdata' | 'rendered' | 'thumbnail'
 
 type WadoRsArgs = {
   studyInstanceUid: string;
-  seriesInstanceUid: string;
-  sopInstanceUid: string;
+  seriesInstanceUid?: string;
+  sopInstanceUid?: string;
+  dataFormat?: DataFormat;
+  frame?: number | number[];
 };
 type WadoRsResponse = {
   contentType: string;
   buffer: Buffer;
 };
-export async function doWadoRs({ studyInstanceUid, seriesInstanceUid, sopInstanceUid }: WadoRsArgs): Promise<WadoRsResponse> {
-  const logger = LoggerSingleton.Instance;
-  const storagePath = config.get(ConfParams.STORAGE_PATH) as string;
-  const studyPath = path.join(storagePath, studyInstanceUid);
-  const pathname = path.join(studyPath, sopInstanceUid);
+type QidoResponse = {
+  [key: string]: {
+    Value: string[]
+    vr: string
+  }
+}
+
+const term = '\r\n';
 
-  const exists = await fileExists(pathname);
+/**
+ * This function uses DCMTK to convert the given DICOM file to JPEG format
+ * It will try twice, once without frames, the other with all frames.
+ * If both fail it will throw the error returned by DCMTK.
+ * If no resulting JPEG can be found, then it will return undefined.
+ * 
+ * @param filepath Path to the file to convert
+ * @param [asThumbnail=false] Return as thumbnail
+ */
+async function convertToJpeg(filepath: string, asThumbnail = false) {
+  try {
+    await execFile(
+      'dcmj2pnm',
+      ['+oj', '+Jq', asThumbnail ? '10' : '100', filepath, `${filepath}.jpg`]
+    );
+  }
+  catch (e) {
+    // Try again but with all frames - if this fails don't catch the error (fail!)
+    await execFile(
+      'dcmj2pnm',
+      ['+oj', '+Jq', asThumbnail ? '10' : '100', '+Fa', filepath, `${filepath}`]
+    );
+  }
+  let filePath = `${filepath}.jpg`;
```

Comment:
could you also not move this to within the try, and the 0.jpg file path in the catch? then only call fileExists once.